client-apis/encrypt.md

```diff
@@ -256,42 +242,28 @@ the [signature calculation](#signature-calculation).
 The encrypted message outputted by this operation MUST have a message body equal
 to the message body calculated in this step.
 
-While there MAY still be plaintext left to encrypt,
-this operation MUST either wait for more plaintext to become available,
-wait for an end to the plaintext to be indicated,
-or perform encryption on the available plaintext to construct frames
-which make up the message body.
-
-Frames MUST be constructed sequentially such that the concatenation of their decryption
-in [sequence number](#.../data-format/message-body.md#sequence-number)
-order is equal to the input plaintext which has been encrypted so far.
-
-If more input plaintext MAY become available
-and there are not enough input plaintext bytes available to create a new
-[regular frame](#../data-format/message-body.md#regular-frame),
-then this operation MUST wait until either enough plaintext bytes become available
-or the caller indicates an end to the plaintext.
-
-If more input plaintext MAY become available
-and there is enough plaintext available to construct a new regular frame,
-then this operation MUST [construct a regular frame](#construct-a-frame)
-with the available plaintext.
-
-If more input plaintext MUST NOT become available
-and there are exactly enough plaintext bytes available to create one regular frame,
-then this operation MUST [construct either a final frame or regular frame](#construct-a-frame)
-with the remaining plaintext.
-If they construct a regular frame, they MUST also construct an empty final frame.
-
-If more input plaintext MUST NOT become available
-and there are not enough input plaintext bytes available to create a new regular frame,
-then this operation MUST [construct a final frame](#construct-a-frame)
-with the remaining plaintext.
-
 If [Plaintext Length Bound](#plaintext-length-bound) was specified on input
-and this operation determines that the plaintext being encrypted has a length greater than this value,
+and this operation determines at any time that the plaintext being encrypted
+has a length greater than this value,
 this operation MUST immediately fail.
 
+Before it is indicated that no more plaintext will be available,
+this operation MUST process as much of the available bytes as possible
+by [constructing regular frames](#construct-a-frame).
+
+When it is indicated that no more plaintext will be available,
+this operation MUST perform the following until all available plaintext is processed:
+
+- If there are enough input plaintext bytes available to create a new regular frame,
+  then this operation MUST [construct a regular frame](#construct-a-frame).
+  using the available plaintext.
+- If there are exactly enough plaintext bytes available to create one regular frame,
+  then this operation MUST [construct either a final frame or regular frame](#construct-a-frame)
+  with the remaining plaintext.
+  If they construct a regular frame, they MUST also construct an empty final frame.
```

Comment:
There's actually a bug here with the edge case of an empty plaintext: technically the while loop immediately terminates and you don't require an empty final frame.

It may be better to move this line to after the loop and state "If a final frame has not yet been constructed, they MUST construct an empty final frame"